### DISCUSSION TITLE
Does supabase.auth.getUser() refresh token with @supabase/ssr?

### DISCUSSION THREAD

**Discussion by hawkcookie** [supabase/supabase#22362] [Questions] [answered]
I use Remix * Cloudflare Pages , "@supabase/ssr": "^0.1.0".  I call this method for authetication every time users change routes. But, in production, we often came across /auth/v1/users returns 401 before refresh token may not be expired.

`Since the unencoded session data is retrieved from the local storage medium, do not rely on it as a source of trusted data on the server. It could be tampered with by  the sender. If you need verified, trustworthy user data, call getUser instead.`  is in the docs.

If I want to keep users logged in until the refresh token expires, should I call getSession before getUser?

```ts
export const verifyUser = async (context: AppLoadContext, request: Request) => {
  const { SUPABASE_URL, SUPABASE_ANON_KEY } = context.env;
  const headers = new Headers();

  const supabase = createSupabaseServerClient(SUPABASE_URL, SUPABASE_ANON_KEY, request, headers);
  const {
    data: { user },
    error,
  } = await supabase.auth.getUser();
  if (error) {
    if (error.name === "AuthApiError") {
      const { error: signOutError } = await supabase.auth.signOut();
      if (signOutError) throw Error(signOutError.message);

      throw redirect("/login", { headers });
    } else {
      throw ErrorResponseInLoader(error);
    }
  }
  if (user === null) {
    throw redirect("/login", { headers });
  }
  return { supabase, user, headers };
};
```

**Comment by charislam** [comment]
Is there an error message associated with those 401 responses?

`getUser` shouldn't be 401ing if the refresh token isn't expired; it should just return you the current user. And calling `getSession` before it unfortunately wouldn't accomplish anything. I'm guessing the error is somewhere else, but hard to tell from the provided code snippet.

**Comment by hawkcookie** [comment]
@charislam thanks for your reply

When I check auth logs, I found randomly `/auth/v1/users` returns 401. and it caused `401: invalid claim: missing sub claim.`

```
Log Timestamp (UTC)
2024-04-01T02:19:05.000Z

Log Event Message
{"component":"api","error":"401: invalid claim: missing sub claim","level":"info","method":"GET","msg":"401: invalid claim: missing sub claim","path":"/user","referer":"XXXXX","remote_addr":"2a06:98c0:3600::103","time":"2024-04-01T02:19:05Z","timestamp":"2024-04-01T02:19:05Z"}

Log Metadata
[
  {
    "message": null,
    "timestamp": "2024-04-01T02:19:05Z",
    "__MONOTONIC_TIMESTAMP": null,
    "CODE_FUNC": null,
    "instance_id": null,
    "status": null,
    "_CMDLINE": null,
    "method": "GET",
    "_SYSTEMD_CGROUP": null,
    "CODE_FILE": null,
    "EXECUTABLE": null,
    "_EXE": null,
    "UNIT": null,
    "level": "info",
    "_COMM": null,
    "duration": null,
    "issuer": null,
    "_LINE_BREAK": null,
    "_SOURCE_REALTIME_TIMESTAMP": null,
    "msg": "401: invalid claim: missing sub claim",
    "action": null,
    "login_method": null,
    "_UID": null,
    "host": "XXXXXXX",
    "PRIORITY": null,
    "_CAP_EFFECTIVE": null,
    "_PID": null,
    "INVOCATION_ID": null,
    "_SYSTEMD_UNIT": null,
    "source_type": null,
    "SYSLOG_FACILITY": null,
    "request_id": null,
    "CODE_LINE": null,
    "path": "/user",
    "component": "api",
    "project": null,
    "user_id": null,
    "auth_event": [],
    "args": [],
    "referer": "XXXXXXX",
    "factor_id": null,
    "provider": null,
    "client_id": null,
    "remote_addr": "2a06:98c0:3600::103",
    "_SYSTEMD_SLICE": null,
    "_SYSTEMD_INVOCATION_ID": null,
    "header": null,
    "_MACHINE_ID": null,
    "_AUDIT_LOGINUID": null,
    "_TRANSPORT": null,
    "_SELINUX_CONTEXT": null,
    "MESSAGE_ID": null,
    "__REALTIME_TIMESTAMP": null,
    "metadata": [],
    "_STREAM_ID": null,
    "metering": null,
    "time": null,
    "_GID": null,
    "_BOOT_ID": null,
    "SYSLOG_IDENTIFIER": null,
    "_AUDIT_SESSION": null,
    "error": "401: invalid claim: missing sub claim"
  }
]

```


I don't understand why this phenomenon is occurring. The documentation states that calling getSession() should refresh the access token using the refresh token, and I thought that just calling getUser would result in a 401 when the 1-hour validity of the access token expires because the user's token wouldn't be refreshed. In any case, what I want to solve is for logged-in users to remain logged in for a relatively long period.

**Comment by charislam** [comment]
hmm looks like a problem with JWT refresh, as you say

what does your `createSupabaseServerClient` look like?
and how are you handling token refresh?

**Reply by hawkcookie** [reply]
Because I use @supabase/ssr,  I have implemented it based on the documentation as follows.


```ts
const createSupabaseServerClient = (
  url: string,
  key: string,
  request: Request,
  headers: Headers,
) => {
  const cookies = parse(request.headers.get("Cookie") ?? "");
  return createServerClient<Database>(url, key, {
    cookies: {
      get(key) {
        return cookies[key];
      },
      set(key, value, options) {
        headers.append("Set-Cookie", serialize(key, value, options));
      },
      remove(key, options) {
        headers.append("Set-Cookie", serialize(key, "", options));
      },
    },
    cookieOptions: {
      secure: process.env.NODE_ENV === "production",
    },
  });
};

export default createSupabaseServerClient;

```

**Comment by charislam** [comment]
And this is only happening in prod, you say? So you only have the logs for debugging? Do you know exactly which line of code triggers the error?

What I suspect might be happening is that in your app, you call `getUser` somewhere without gating behind a valid sign-in.

**Comment by hawkcookie** [comment]
@charislam I am using the getUser function before displaying the login screen to prevent the logged-in user from being redirected to the login screen.

```ts
export const loader = async ({ context, request }: LoaderFunctionArgs) => {
  const headers = new Headers();
  const { SUPABASE_URL, SUPABASE_ANON_KEY } = context.env;
  const supabase = createSupabaseServerClient(SUPABASE_URL, SUPABASE_ANON_KEY, request, headers);

  const {
    data: { user },
    error,
  } = await supabase.auth.getUser();
  if (error) {
    if (error.name === "AuthApiError") {
      const { error: signOutError } = await supabase.auth.signOut();
      if (signOutError) throw ErrorResponseInLoader(signOutError);

      return json({}, { headers });
    } else {
      throw ErrorResponseInLoader(error);
    }
  }
  if (user === null) return null;

  return redirectWithInfo("/account", "既にログイン済みです", { headers });
};
```

**Comment by charislam** [comment]
Ah, that makes perfect sense then

So you were right and I misunderstood the use case, you could check for a valid session before calling `getUser`, and if there's no valid session, return even earlier

**Comment by hawkcookie** [comment]
@charislam Thanks. Sorry to return to this question about GitHub discussions, in the end, when using `@supabase/ssr` to call `getUser`, is the token automatically refreshed? The documentation for `getUser` doesn't mention this point, and for `getSession`, it is noted that the token is automatically refreshed. However, when actually running the application in a production environment, using `getUser` alone seems to maintain the user's logged-in state, and the token appears to be refreshed.

**Comment by charislam** [comment, marked as answer]
Yes it does! Under the hood `getUser` gets the session using the same methods as `getSession` ([code here](https://github.com/supabase/auth-js/blob/bd91e72824ceb075f6fca7ae25bf9f066e6508d2/src/GoTrueClient.ts#L1195-L1207)), which includes refresh logic, before going on to call the Supabase backend for user info.

**Reply by hawkcookie** [reply]
I am very grateful for your support. Thank you very much!
